xilem_web/src/as_template.rs

```diff
@@ -0,0 +1,132 @@
+            el.node.apply_props(&mut el.props);
+            if !is_already_hydrating {
+                ctx.disable_hydration();
+            }
+            let pod_mut = PodMut::new(&mut el.node, &mut el.props, None, false);
+            self.0.rebuild(prev, &mut state, ctx, pod_mut);
+
+            (el, state)
+        } else {
+            let (element, state) = self.0.build(ctx);
+
+            let template: web_sys::Node = element
+                .node
+                .as_ref()
+                .clone_node_with_deep(true)
+                .unwrap_throw();
+
+            ctx.templates.insert(type_id, (template, self.0.clone()));
+            (element, state)
+        };
+        let state = AsTemplateState {
+            view_state,
+            dirty: false,
+        };
+        (element, state)
+    }
+
+    fn rebuild<'el>(
+        &self,
+        prev: &Self,
+        view_state: &mut Self::ViewState,
+        ctx: &mut ViewCtx,
+        element: xilem_core::Mut<'el, Self::Element>,
+    ) -> xilem_core::Mut<'el, Self::Element> {
+        if core::mem::take(&mut view_state.dirty) || !Rc::ptr_eq(&self.0, &prev.0) {
+            self.0
+                .deref()
+                .rebuild(&prev.0, &mut view_state.view_state, ctx, element)
+        } else {
+            // If this is the same value, or no rebuild was forced, there's no need to rebuild
+            element
+        }
+    }
+
+    fn teardown(
+        &self,
+        view_state: &mut Self::ViewState,
+        ctx: &mut ViewCtx,
+        element: xilem_core::Mut<'_, Self::Element>,
+    ) {
+        self.0.teardown(&mut view_state.view_state, ctx, element);
+    }
+
+    fn message(
+        &self,
+        view_state: &mut Self::ViewState,
+        id_path: &[xilem_core::ViewId],
+        message: DynMessage,
+        app_state: &mut State,
+    ) -> xilem_core::MessageResult<Action, DynMessage> {
+        let message_result =
+            self.0
+                .deref()
+                .message(&mut view_state.view_state, id_path, message, app_state);
+        if matches!(message_result, MessageResult::RequestRebuild) {
+            view_state.dirty = true;
+        }
+        message_result
+    }
+}
+
+/// This view creates an internally cached deep-clone of the underlying DOM node.
+/// When the inner view is created again, this will be done more efficiently.
+/// It's recommended to use this as wrapper, when it's expected that the inner `view` will be created a lot, for example in a long list
+///
+/// Additionally it supports memoization when the given `view` is an [`Rc<impl DomView>`].
+///
+/// # Examples
+///
+/// ```
+/// use xilem_web::{as_template, elements::html};
+///
+/// # use xilem_web::interfaces::Element;
+/// # fn component() -> impl Element<()> {
+/// as_template(html::p("This will likely created a lot"))
+/// # }
+/// ```
+pub fn as_template<State, Action, E>(view: impl Into<Rc<E>>) -> AsTemplate<E>
```

Comment:
Isn't this more a `into_template`?

Reply:
Hmm not sure, it's not *only* being used as template, the initial instance of this is cached, any follow-up is deep-cloned and rebuild by the new values.

I mean the name could be either `instanced`, `as_template`, `into_template` or something else (I had `templated` as well, hmm finding good names for views is hard).

(Thanks for the review btw. will address the suggestions soon)

Reply:
I think `templated` is actually even better than `as_template` and `into_template`.

> hmm finding good names for views is hard.

Yes, one of the most difficult things when it comes to public APIs, which should ideally not change and be intuitively understandable for everyone.
Ideally, all views should also be consistent in their naming.

Reply:
I've renamed it now `templated` seems to be a good compromise.

Reply:
Nice! :+1: